Added unique uncrafting recipes

jodlodi opened the pull request:
Added unique crafting recipes. They can only be uncrafted and can not be crafted. The input is an ingredient with a count and the output is a list of up to 9 ingredients. A custom level cost can be set.

These custom recipes work with JEI the same way as normal uncrafting recipes do.

An example .json recipe is:
{
  "type": "twilightforest:uncrafting",
  "cost": 1,
  "ingredient": [
    {
      "item": "twilightforest:zombie_scepter"
    },
    {
      "item": "twilightforest:twilight_scepter"
    },
    {
      "item": "twilightforest:lifedrain_scepter"
    },
    {
      "item": "twilightforest:fortification_scepter"
    }
  ],
  "count": 1,
  "pattern": [
    "   ",
    " 1 ",
    "0  "
  ],
  "key": {
    "0": {
      "item": "minecraft:bone"
    },
    "1": {
      "item": "minecraft:ender_pearl"
    }
  }
}

jodlodi committed:
JEI uncrafting support

Added a UncraftingCategory class for JEI, a gui element for uncrafting in JEI, implemented it into the JEI class and added the required string to the lang file

jodlodi committed:
Fixed rendering error, fixed empty recipes

jodlodi committed:
Final patch, banned uncrafting items fix

Uncrafting JEI now properly deals with banned items and doesn't show them in recipes.

jodlodi committed:
Last last update

Made it so it wont show recipes that have every single ingredient banned.
It looked weird.

jodlodi committed:
Uncrafting recipe

Added custom uncrafting recipe type

jodlodi committed:
Merge branch '1.17.x' of https://github.com/TeamTwilight/twilightforest into 1.17.x

jodlodi committed:
Merge branch '1.17.x' of https://github.com/jodlodi/twilightforest into 1.17.x

Tamaized commented on src/main/java/twilightforest/item/recipe/UncraftingRecipe.java:
```diff
@@ -0,0 +1,260 @@
+
+    @Override //We have to implement this method, can't really be used since we have multiple outputs
+    public ItemStack assemble(CraftingContainer pContainer) {
+        return null;
```
Should probably use ItemStack.EMPTY Incase a modder calls this or something

jodlodi committed:
Made assemble() return ItemStack.EMPTY instead of null

jodlodi commented:
I have one last fix to push

jodlodi committed:
Uncrafting shift-click fix

While in the uncrafting menu, shift-clicking on an item in your inventory now transfers it into the uncrafting input slot. Shift clicking on an item in the uncrafting result matrix now properly only gives the player the item available without depleting the rest of the items in the uncrafting input slot.

Additionally, custom uncrafting recipe logic is slightly changed to probably be better.

I can't find any more bugs.

jodlodi committed:
Fixed a comment :)

so that it's correct
